// babble.js/core/stage.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// imports
var PIXI = require('pixi.js');
window.PIXI = PIXI;
window.PIXI["default"] = PIXI;
var timer = require('pixi-timer');
var Puppet = require('./puppet');
var path = require('path');
var trim = require('./../util/trimCanvas');

// Constants
var MOVE_DURATION = 0.75; // in seconds

// Aliases
var BaseTextureCache = PIXI.utils.BaseTextureCache,
    Container = PIXI.Container,
    Sprite = PIXI.Sprite,
    Texture = PIXI.Texture,
    TextureCache = PIXI.utils.TextureCache,
    autoDetectRenderer = PIXI.autoDetectRenderer,
    loader = PIXI.loader,
    Rectangle = PIXI.Rectangle,
    ticker = PIXI.ticker;

/**
 * @class
 */

var Stage = function () {

    /**
     * @param {string} element - the id of the DOM element to append the stage to
     * @param {Object} project - object with information on the assets, puppets, and stage settings
     * @param {Object[]} assets - array of assets
     * @param {string} assetsPath - path to the assets folder
     * @param {requestCallback} callback - function to be called after assets are loaded
     * @param {Object} [status] - object for logging stuff
     * @param {boolean} [enabled=true] - whether or not it should start updating from the start
     */
    function Stage(element, project, assets, assetsPath, callback, status, enabled) {
        var _this = this;

        _classCallCheck(this, Stage);

        this.project = project;
        this.assets = assets;
        this.assetsPath = assetsPath;
        this.status = status;
        this.MOVE_DURATION = MOVE_DURATION;
        this.enabled = enabled === undefined ? true : enabled;

        // Create some basic objects
        this.stage = new Container();
        this.puppetStage = new Container();
        this.stage.addChild(this.puppetStage);
        this.renderer = autoDetectRenderer(1, 1, { transparent: true });
        this.screen = document.getElementById(element);
        this.screen.appendChild(this.renderer.view);

        this.lastFrame = new Date();
        this.puppets = [];
        this.listeners = [];

        // Make the game fit the entire window
        this.renderer.view.style.position = "absolute";
        this.renderer.view.style.display = "block";

        // Load Assets
        if (loader.loading) {
            this.resize();
            if (callback) requestAnimationFrame(function () {
                callback(_this);
            });
            this.gameLoop();
            return;
        }
        var texturesToLoad = false;
        Object.values(assets).forEach(function (asset) {
            if (!TextureCache[path.join(assetsPath, asset.location)]) {
                loader.add(path.join(assetsPath, asset.location));
                texturesToLoad = true;
            }
        });
        var stage = this;
        if (texturesToLoad) {
            loader.onComplete.once(function () {
                stage.resize();
                if (callback) requestAnimationFrame(function () {
                    callback(stage);
                });
                stage.gameLoop();
            });
            loader.load();
        } else {
            loader.load();
            stage.resize();
            if (callback) requestAnimationFrame(function () {
                callback(stage);
            });
            stage.gameLoop();
        }
    }

    _createClass(Stage, [{
        key: 'registerPuppetListener',
        value: function registerPuppetListener(event, callback) {
            this.listeners.push({ "event": event, "callback": callback });
            this.puppets.forEach(function (p) {
                return p.container.on(event, callback);
            });
        }
    }, {
        key: 'addAsset',
        value: function addAsset(id, asset, callback) {
            this.assets[id] = asset;
            var date = new Date();
            TextureCache[path.join(this.assetsPath, asset.location)] = Texture.fromImage(path.join(this.assetsPath, asset.location + "?random=" + date));
            BaseTextureCache[path.join(this.assetsPath, asset.location)] = BaseTextureCache[path.join(this.assetsPath, asset.location + "?random=" + date)];
            if (callback) TextureCache[path.join(this.assetsPath, asset.location)].baseTexture.on('loaded', callback);
        }
    }, {
        key: 'reloadAssets',
        value: function reloadAssets(callback) {
            var _this2 = this;

            Object.keys(TextureCache).forEach(function (k) {
                if (TextureCache[k]) TextureCache[k].destroy(true);
            });

            // Load Assets
            Object.values(this.assets).forEach(function (a) {
                if (!TextureCache[path.join(_this2.assetsPath, a.location)]) TextureCache[path.join(_this2.assetsPath, a.location)] = Texture.fromImage(path.join(_this2.assetsPath, a.location));
            });
            var stage = this;
            var onLoad = function onLoad() {
                if (!Object.values(BaseTextureCache).some(function (a) {
                    return a.isLoading;
                })) {
                    callback(stage);
                    ticker.shared.remove(onLoad);
                }
            };

            this.reloadPuppets();
            if (callback) {
                ticker.shared.add(onLoad);
            }
        }
    }, {
        key: 'updateAsset',
        value: function updateAsset(id) {
            var stage = this;
            var callback = function callback(asset, sprite, layer, emote) {
                var parent = sprite.parent;
                var index = parent.getChildIndex(sprite);
                var newAsset = stage.getAsset(asset, layer, emote);
                parent.removeChildAt(index);
                parent.addChildAt(newAsset, index);
            };
            this.puppets.forEach(function (p) {
                return p.applyToAsset(id, callback);
            });
        }
    }, {
        key: 'reloadPuppets',
        value: function reloadPuppets() {
            var _this3 = this;

            this.puppets.forEach(function (p) {
                return _this3.setPuppet(p.id, _this3.createPuppet(p.puppet));
            });
        }
    }, {
        key: 'reattach',
        value: function reattach(element) {
            this.screen = document.getElementById(element);
            this.screen.appendChild(this.renderer.view);
            this.resize();
        }
    }, {
        key: 'resize',
        value: function resize(e, width, height) {
            var rect = this.screen.getBoundingClientRect();
            this.bounds = {
                width: width || rect.width,
                height: height || rect.height
            };
            this.renderer.resize(this.bounds.width, this.bounds.height);
            this.slotWidth = this.bounds.width / this.project.numCharacters;
            if (this.slotWidth < 400) {
                this.puppetStage.scale.x = this.puppetStage.scale.y = this.slotWidth / 400;
                this.slotWidth = 400;
            } else this.puppetStage.scale.x = this.puppetStage.scale.y = 1;

            this.puppets.forEach(function (p) {
                return p.updatePosition();
            });
        }
    }, {
        key: 'createPuppet',
        value: function createPuppet(puppet) {
            return new Puppet(this, puppet, -1);
        }
    }, {
        key: 'addPuppet',
        value: function addPuppet(puppet, id) {
            var newPuppet = new Puppet(this, puppet, id);
            this.puppets.push(newPuppet);
            this.puppetStage.addChild(newPuppet.container);
            this.listeners.forEach(function (l) {
                return newPuppet.container.on(l.event, l.callback);
            });
            newPuppet.updatePosition();
            return newPuppet;
        }
    }, {
        key: 'removePuppet',
        value: function removePuppet(id) {
            var puppet = this.puppets.find(function (p) {
                return p.id == id;
            });
            if (puppet) {
                this.puppets.splice(this.puppets.indexOf(puppet), 1);
                this.puppetStage.removeChild(puppet.container);
            }
        }
    }, {
        key: 'clearPuppets',
        value: function clearPuppets() {
            while (this.puppets.length !== 0) {
                this.puppetStage.removeChild(this.puppets[0].container);
                this.puppets.splice(0, 1);
            }
        }
    }, {
        key: 'banishPuppets',
        value: function banishPuppets() {
            var _this4 = this;

            this.puppets.forEach(function (puppet) {
                if (puppet.target > _this4.project.numCharacters / 2) {
                    puppet.target = _this4.project.numCharacters + 1;
                    puppet.facingLeft = false;
                    puppet.container.scale.x = _this4.project.puppetScale || 1;
                } else {
                    puppet.target = 0;
                    puppet.facingLeft = true;
                    puppet.container.scale.x = -1 * (_this4.project.puppetScale || 1);
                }
            });
        }
    }, {
        key: 'getPuppet',
        value: function getPuppet(id) {
            return this.puppets.find(function (p) {
                return p.id == id;
            });
        }
    }, {
        key: 'setPuppet',
        value: function setPuppet(id, newPuppet) {
            var oldPuppet = this.getPuppet(id);
            newPuppet.changeEmote(oldPuppet.emote);
            newPuppet.id = newPuppet.container.id = oldPuppet.id;
            newPuppet.position = oldPuppet.position;
            newPuppet.target = oldPuppet.target;
            newPuppet.facingLeft = oldPuppet.facingLeft;
            newPuppet.babbling = oldPuppet.babbling;
            this.listeners.forEach(function (l) {
                return newPuppet.container.on(l.event, l.callback);
            });

            this.puppets[this.puppets.indexOf(oldPuppet)] = newPuppet;
            this.puppetStage.removeChild(oldPuppet.container);
            this.puppetStage.addChild(newPuppet.container);
            this.resize();

            return newPuppet;
        }
    }, {
        key: 'getThumbnail',
        value: function getThumbnail() {
            this.renderer.render(this.stage);
            try {
                return trim(this.renderer.plugins.extract.canvas(this.stage)).canvas.toDataURL().replace(/^data:image\/\w+;base64,/, "");
            } catch (e) {
                this.status.error("Failed to generate thumbnail", e);
                return null;
            }
        }
    }, {
        key: 'gameLoop',
        value: function gameLoop() {
            var thisFrame = new Date();
            var delta = thisFrame - this.lastFrame;
            this.lastFrame = thisFrame;

            requestAnimationFrame(this.gameLoop.bind(this));
            if (this.enabled) this.update(delta);
        }
    }, {
        key: 'getAsset',
        value: function getAsset(container, asset) {
            var sprite = void 0;
            if (this.assets[asset.id]) {
                var assetData = this.assets[asset.id];
                if (assetData.type === "animated") {
                    var base = BaseTextureCache[path.join(this.assetsPath, assetData.location)];
                    var textures = [];
                    var width = base.width / assetData.cols;
                    var height = base.height / assetData.rows;
                    for (var i = 0; i < assetData.numFrames; i++) {
                        if (i % assetData.cols * width + width > base.width || Math.floor(i / assetData.cols) * height + height > base.height) continue;
                        var rect = new Rectangle(i % assetData.cols * width, Math.floor(i / assetData.cols) * height, width, height);
                        textures.push(new Texture(base, rect));
                    }
                    sprite = new PIXI.extras.AnimatedSprite(textures);
                    sprite.animationSpeed = 20 / assetData.delay;
                    sprite.play();
                } else sprite = new Sprite(TextureCache[path.join(this.assetsPath, assetData.location)]);
            } else {
                sprite = new Sprite();
                if (this.status) this.status.log("Unable to load asset \"" + asset.id + "\"", 5, 2);
            }
            sprite.scale.set(asset.scaleX, asset.scaleY);
            sprite.anchor.set(.5, .5);
            container.addChild(sprite);
            container.x = asset.x;
            container.y = asset.y;
            container.rotation = asset.rotation;
            container.asset = asset;
            return container;
        }
    }, {
        key: 'update',
        value: function update(delta) {
            var _this5 = this;

            this.puppets.forEach(function (puppet) {
                // Movement animations
                // I've tried to emulate what puppet pals does as closely as possible
                // But frankly it's difficult to tell
                if (puppet.target != puppet.position || puppet.movingAnim !== 0) {
                    // Whether its going left or right
                    if (puppet.direction === 0 && puppet.target != puppet.position) puppet.direction = puppet.target > puppet.position ? 1 : -1;
                    // Update how far into the animation we are
                    puppet.movingAnim += delta / (1000 * MOVE_DURATION);

                    // We want to do a bit of animation when they arrive at the target slot. 
                    //  in order to do that we have part of the animation (0 - .6) be for each slot
                    //  and the rest (.6 - 1) only plays at the destination slot
                    if (puppet.movingAnim >= 0.6 && puppet.movingAnim - delta / (1000 * MOVE_DURATION) < 0.6) {
                        // Once we pass .6, update our new slot position
                        puppet.position += puppet.direction;
                        puppet.direction = 0;
                        // If we're not at the final slot yet, reset the animation
                        if (puppet.position != puppet.target) puppet.movingAnim = 0;else puppet.container.scale.x = (puppet.facingLeft ? -1 : 1) * (_this5.project.puppetScale || 1);
                    } else if (puppet.movingAnim >= 1) puppet.movingAnim = 0;else if (puppet.movingAnim < 0.6) puppet.container.scale.x = puppet.direction * (_this5.project.puppetScale || 1);

                    // Scale in a sin formation such that it does 3 half circles per slot, plus 2 more at the end
                    puppet.container.scale.y = (1 + Math.sin((1 + puppet.movingAnim * 5) * Math.PI) / 40) * (_this5.project.puppetScale || 1);
                    // Update y value so it doesn't leave the bottom of the screen while bouncing
                    puppet.container.y = _this5.bounds.height / _this5.puppetStage.scale.y;
                    // Linearly move across the slot, unless we're in the (.6 - 1) part of the animation, and ensure we're off screen even when the puppets are large
                    var interpolation = Math.min(1, puppet.movingAnim / 0.6);
                    var pos = puppet.position % (_this5.project.numCharacters + 1);
                    if (pos < 0) pos += _this5.project.numCharacters + 1;
                    var start = pos == 0 ? puppet.direction === 1 ? -Math.abs(puppet.container.width) : // Starting on left edge of screen
                    _this5.project.numCharacters * _this5.slotWidth + Math.abs(puppet.container.width) : // Starting on right edge of screen
                    (pos - 0.5) * _this5.slotWidth; // Ending on screen
                    pos += puppet.direction;
                    if (pos < 0) pos += _this5.project.numCharacters + 1;
                    var end = pos <= 0 ? -Math.abs(puppet.container.width) : // Starting left of screen
                    pos >= _this5.project.numCharacters + 1 ? _this5.project.numCharacters * _this5.slotWidth + Math.abs(puppet.container.width) : // Starting right of screen
                    (pos - 0.5) * _this5.slotWidth; // Ending on screen
                    puppet.container.x = interpolation === 1 ? start : start + (end - start) * interpolation;
                }
                if (puppet.babbling) {
                    // Update how long each face part has been on display
                    puppet.eyesAnim += delta;
                    puppet.mouthAnim += delta;

                    // Update eyes
                    if (puppet.eyesAnim >= puppet.eyesDuration && (puppet.emote === '0' || !puppet.emotes[puppet.emote])) {
                        puppet.updateEyeBabble();
                    }

                    // Update mouth
                    if (puppet.mouthAnim >= puppet.mouthDuration) {
                        puppet.updateMouthBabble();
                    }
                }
                // Update DeadbonesStyle Babbling
                // I'm not sure what Puppet Pals does, but I'm pretty sure this isn't it
                // But I think this looks "close enough", and probably the best I'm going
                // to get without Rob actually telling people how Puppet Pals does it
                if (puppet.deadbonesStyle && (puppet.babbling || puppet.deadbonesDuration !== 0)) {
                    puppet.deadbonesAnim += delta;
                    if (puppet.deadbonesAnim >= puppet.deadbonesDuration) {
                        puppet.deadbonesAnim = 0;
                        if (puppet.babbling) {
                            puppet.deadbonesDuration = 100 + Math.random() * 200;
                            puppet.deadbonesStartY = puppet.deadbonesTargetY;
                            puppet.deadbonesStartRotation = puppet.deadbonesTargetRotation;
                            puppet.head.forEach(function (a) {
                                a.y = a.asset.y + puppet.deadbonesStartY;
                                a.rotation = a.asset.rotation + puppet.deadbonesStartRotation;
                            });
                            puppet.deadbonesTargetY = 10 - Math.random() * 20;
                            puppet.deadbonesTargetRotation = 0.1 - Math.random() * 0.2;
                        } else {
                            puppet.deadbonesDuration = 0;
                            puppet.head.forEach(function (a) {
                                a.y = a.asset.y + puppet.deadbonesTargetY;
                                a.rotation = a.asset.rotation + puppet.deadbonesTargetRotation;
                            });
                        }
                    } else {
                        var percent = puppet.deadbonesAnim / puppet.deadbonesDuration * (puppet.deadbonesAnim / puppet.deadbonesDuration);
                        puppet.head.forEach(function (a) {
                            a.y = a.asset.y + puppet.deadbonesStartY + (puppet.deadbonesTargetY - puppet.deadbonesStartY) * percent;
                            a.rotation = a.asset.rotation + puppet.deadbonesStartRotation + (puppet.deadbonesTargetRotation - puppet.deadbonesStartRotation) * percent;
                        });
                    }
                }
            });
            this.renderer.render(this.stage);
            PIXI.timerManager.update(delta / 1000);
        }
    }]);

    return Stage;
}();

module.exports = Stage;
